File: leetcode/206-reverse-linked-list.cpp

struct ListNode {
  int val;
  ListNode *next;
  ListNode(int x) : val(x), next(nullptr) {}
};

class Solution {
public:
  ListNode *reverseList(ListNode *head) {
    if (!head || !(head->next))
      return head;
    ListNode *node = reverseList(head->next);
    head->next->next = head;
    head->next = nullptr;
    return node;
  }
};

File: leetcode/206-reverse-linked-list.py
class ListNode:
    def __init__(self, x):
        self.val = x
        self.next = None


class Solution:
    def reverseList(self, head):
        prior = None
        current = head

        while current is not None:
            next_node = current.next
            current.next = prior
            prior = current
            current = next_node

        return prior
